### ISSUE TITLE
Weapons which have forced attachments lose their attachments when stacked

### ISSUE THREAD

**Issue by BlackDog86** [X2CommunityCore/X2WOTCCommunityHighlander#1429]
Some weapons (like the Tactical Legacy Pack weapons) do not set a Template.NumUpgradeSlots on the weapon template, and instead, have their attachments added via a custom OnAcquired function, meaning that by design, the attachments cannot be changed or removed. If mods end up creating more than one of these weapons, they become stacked in the inventory.

When this happens, there is a bug wherein the attachments may be removed when switching loadouts. 

Here are the exact steps to reproduce the issue:

1. Console command: giveitem TLE_AssaultRifle_MG 2 
2. Go to soldier loadout and equip one of the stacked TLE rifles

Attchments will be removed from the weapon when equipped (GetItemFromInventory creates a new item state when the item is equipped but this item does not get its upgrades from the onacquired function when the new item state is created). 

From there, it is then possible to permanently remove the attachments by swapping between the stacked weapons and normal ones - If HasUnmodifiedItem in XCGS_HeadquartersXCom reports that the item is unmodified (i.e. in the case of the weapon with the stripped attachments), it will then just increase the quantity of the stripped weapon instead of putting the old one back in the inventory, with its attachments.

Speculation on a fix would be to run the items onacquired function (if it exists) when equipping but I'm unsure whether this would have any adverse effects on other weapons which may have onacquired functions which do weird & wonderful things that you wouldn't want to run every time you equipped the item.

**Comment by Iridar** [comment]
Normally it should be impossible to acquire multiples of these weapons other than with a console command. Is there even a way to trigger the issue via conventionally modded gameplay?

**Comment by Dos5904** [comment]
> Normally it should be impossible to acquire multiples of these weapons other than with a console command. Is there even a way to trigger the issue via conventionally modded gameplay?

Yes, I am currently playing with ChristopherOdd's S9 mod pack - I have received double ups on TLE weapons multiple times. The first one is given as usual with upgrading to mag / plasma weapons, and copies are given as rewards, from memory mine was from a protoype item scanning site or something similar.

**Comment by Iridar** [comment]
I see two ways this can be fixed:
1. Identify weapons that have identical upgrades installed and allow them to stack, and fix whatever behavior that is causing upgrades to break on them.
2. Disallow weapons with upgrades of any kind to stack, which is what vanilla behavior should be, if I'm not mistaken.

**Comment by BlackDog86** [comment]
Just tested without CHL and the bug is still there so it's base-game. Other types of weapons with upgrades of any kind do not stack so these TLE weapons are an outlier from that behaviour. I think this is the easier of the two choices available to us, but it's kind of a compelling argument that if you have two of the same weapons with the same upgrades attached, that they should probably stack as well.

**Comment by Dos5904** [comment]
I guess a happy medium is that the automatic TLE weapon has the attachments, but any subsequent ones don't unless you have the sitrep available? Just not sure if that is possible.

**Comment by Iridar** [comment]
It's possible, but the onus to accomplish that is on mods that make additional TLE weapons possible.

**Comment by BlackDog86** [comment]
One way of doing it without adjustijng too much of the game logic would be to place an OnEquippedFn on these weapons, which fires the OnAcquiredFn - e.g.
`Template.OnEquippedFn = OnTLE3CannonAcquired;`
This would prevent the stripping of the attachments when the item is equipped (it's what Ashlynne's Warlock mod does with its weapon).

Having said that, the HasBeenModified() function of XCGS_Item is wrong, and returns false for a lot of items with attachments - this should really be adjusted to remove the reliance on GetNumUpgradeSlots to account for weapons which have fixed upgrades like these.

**Comment by BlackDog86** [comment]
Simply removing the GetNumUpgradeSlots() check from HasBeenModified() is enough to make that function behave as you would expect & subsequently obtained versions of TLE weapons which are not created in multiples by the console command, will not be stacked. It has the slgihtly unfortunate downside of unstacking all modified weapons in debug strategy start and if you console command in >1 of them, every time you equip one, one will disappear until there's only one left.

**Comment by faanlez** [comment]
My two cents is that there is less chance of things breaking if the default functionality is that weapons with attachments will never stack. This is what the base game and mods would probably expect as well. It is an extremely minor issue that identically upgraded weapons wouldn't stack. Even when mods like Prototype Armoury can give you more than one TLE weapon, you wouldn't get all that many anyway.

**Comment by Larryturbo** [comment]
I've never run any mods that give you multiples of the DLC weapons making it hard for me to rationalize all the proposed edits on a system that has always behaved as I expected it to. If a form of the changes in the pull request makes it to Highlander, it should affect the small number of unique weapons (and the mods that give you more of them), while passing unnoticed for people like me.
